numcodecs/gzip.py

```diff
@@ -28,22 +25,9 @@ def __init__(self, level=1):
 
     def encode(self, buf):
 
-        # deal with lack of buffer support for datetime64 and timedelta64
-        buf = handle_datetime(buf)
-
-        if isinstance(buf, np.ndarray):
-
-            # cannot compress object array
-            if buf.dtype == object:
-                raise ValueError('cannot encode object array')
-
-            # if numpy array, can only handle C contiguous directly
-            if not buf.flags.c_contiguous:
-                buf = buf.tobytes(order='A')
-
-        if PY2:  # pragma: py3 no cover
-            # ensure bytes, PY2 cannot handle things like bytearray
-            buf = buffer_tobytes(buf)
+        # normalise inputs
+        buf = ensure_contiguous_ndarray(buf)
+        buf = memoryview(buf)  # needed on PY2 for unknown reasons
```

Comment:
Had run into this myself as well, but then started wondering why it was an issue.

Turns out discussion in [Python issue 7471]( https://bugs.python.org/issue7471 ) generated [a patch]( https://github.com/python/cpython/commit/673ddf9907898d4483a2622699f588f0b58a5226 ) that made `GzipFile` a subclass of `io.BufferedIOBase`. Part of the patch included an [explicit check for `memoryview` instances]( https://github.com/python/cpython/commit/673ddf9907898d4483a2622699f588f0b58a5226#diff-90c90c977f09602928f0518fc849ca9fR213 ), which [it converted to `bytes`]( https://github.com/python/cpython/commit/673ddf9907898d4483a2622699f588f0b58a5226#diff-90c90c977f09602928f0518fc849ca9fR214 ). Unfortunately there is not much detail as to why this conversion was needed in that discussion.

A follow-up in [Python issue 23688]( https://bugs.python.org/issue23688 ), which attempts to use `memoryview`s without converting to `bytes` [clarifies this point]( https://bugs.python.org/issue23688#msg238307 ). In particular, [`len`]( https://github.com/python/cpython/commit/673ddf9907898d4483a2622699f588f0b58a5226#diff-90c90c977f09602928f0518fc849ca9fR217 ) is used to determine the number of bytes in the data. However this result will be wrong if the data is not of `char` type. This was [fixed]( https://github.com/python/cpython/commit/bca63b362d23f154a5ed7fe43e4146977bba181e ) in Python 3.5+ to actually check `nbytes`, but [that functionality does not exist in Python 2.7]( https://bugs.python.org/issue23688#msg239034 ); so, it was not backported.

Though I think we can do better as we use NumPy. Namely we can change this line to `buf = buf.view('u1')` instead of converting to a `memoryview`, which should avoid the copy and get the correct behavior. We can limit this to Python 2 if that is preferred.

Reply:
You're write, viewing as 'u1' does work on PY2, made this change in latest commit.

Reply:
Adding a link back to this from the code in PR ( https://github.com/zarr-developers/numcodecs/pull/133 ) lest we wonder what this workaround is doing. 😄